galsim/utilities.py

```diff
@@ -330,34 +349,137 @@ def thin_tabulated_values(x, f, rel_err=1.e-4, preserve_range=False):
         # Nothing to do
         return x,f
 
+    total_integ = np.trapz(abs(f), x)
+    thresh = total_integ * rel_err
+
+    if not preserve_range:
+        errleft = 0.5 * (abs(f[1])+abs(f[0])) * (x[1]-x[0])
+        errright = 0.5 * (abs(f[-1])+abs(f[-2])) * (x[-1]-x[-2])
+        accumulated_err = 0.0
+        # arbitrarily allocate half of error budget to trimming the endpoints.
+        # while at least one endpoint can be trimmed:
+        while accumulated_err+min(errleft, errright) < thresh * 0.5:
+            # if either endpoint could be trimmed, then trim the one leading to the smaller error,
+            # under the assumption that this is more likely to lead to trimming a larger total
+            # number of points.
+            if accumulated_err+max(errleft, errright) < thresh * 0.5:
+                if errleft < errright:
+                    x, f = x[1:], f[1:]
+                    accumulated_err += errleft
+                    errleft = 0.5 * (abs(f[1])+abs(f[0])) * (x[1]-x[0])
+                else:
+                    x, f = x[:-1], f[:-1]
+                    accumulated_err += errright
+                    errright = 0.5 * (abs(f[-1])+abs(f[-2])) * (x[-1]-x[-2])
+            elif accumulated_err+errleft < thresh * 0.5:
+                x, f = x[1:], f[1:]
+                accumulated_err += errleft
+                errleft = 0.5 * (abs(f[1])+abs(f[0])) * (x[1]-x[0])
+            else:  # accumulated_err+effright < thresh * 0.5:
+                x, f = x[:-1], f[:-1]
+                accumulated_err += errright
+                errright = 0.5 * (abs(f[-1])+abs(f[-2])) * (x[-1]-x[-2])
+        thresh -= accumulated_err  # in case we didn't use all of thresh*0.5
+
+    # Check again for noop after trimming endpoints.
+    if len(x) <= 2:
+        return x,f
+
+    # Thin interior points.  Start with no interior points and then greedily add them back in one at
+    # a time until relative error goal is met.
+    splitpoints = [0, len(x)-1]  # Start with single interval
+    m = (f[-1]-f[0])/(x[-1]-x[0]) # slope
+    fapprox = f[0]+m*(x-x[0])
+    errs = [np.trapz(np.abs(fapprox-f), x)]  # Current error for each interval
+    while sum(errs) > thresh:
+        # Find the worst current interval
+        index = np.argmax(errs)
+        # Optimally split that interval
+        left = splitpoints[index]
+        right = splitpoints[index+1]
+        i, (errleft, errright) = _lin_approx_split(x[left:right+1], f[left:right+1])
+        # Update splitpoints and errs
+        splitpoints.insert(index+1, i+left)
+        del errs[index]
+        errs.insert(index, errright)
+        errs.insert(index, errleft)
+    return x[splitpoints].tolist(), f[splitpoints].tolist()
```

Comment:
I think the conversion to lists is probably unnecessary.  The other algorithm built up lists, so that's what I output, but I think outputting numpy arrays is just as good.  I think (but didn't check) that all the places we use this would be just as happy with numpy arrays.